File: 58/solve.py
def solve(array, element):
    left = 0
    right = len(array) - 1

    while left < right:
        pivot = (left + right) // 2

        if element == array[pivot]:
            return pivot
        if element == array[right]:
            return right
        if element == array[left]:
            return left

        if element < array[pivot]:
            if element >= array[left] or array[pivot] < array[left]:
                right = pivot - 1
            else:
                left = pivot + 1
        else:
            if element <= array[right] or array[pivot] > array[right]:
                left = pivot + 1
            else:
                right = pivot - 1

    return -1

File: 08/solve.go
package main

func solve(t *TreeNode) int {
	trees, _ := count(t)
	return trees
}

func count(t *TreeNode) (int, bool) {
	if t == nil {
		return 0, false
	}

	if t.left == nil && t.right == nil {
		return 1, true
	}

	leftCount, leftUnival := count(t.left)
	rightCount, rightUnival := count(t.right)

	count := rightCount + leftCount
	unival := leftUnival && rightUnival

	if t.left != nil {
		unival = unival && (t.left.value == t.value)
	}

	if t.right != nil {
		unival = unival && (t.right.value == t.value)
	}

	if unival {
		count++
	}

	return count, unival
}
